pkg/controller/servicemesh/webhook/controller.go

```diff
@@ -0,0 +1,236 @@
+	// Watch secrets in watchedSecretNames
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			return contains(watchedSecretNames, event.Meta.GetName())
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			return contains(watchedSecretNames, event.MetaNew.GetName())
+		},
+		// deletion and generic events don't interest us
+		DeleteFunc: func(event event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(event event.GenericEvent) bool {
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch MutatingWebhookConfigurations
+	err = c.Watch(&source.Kind{Type: &v1beta1.MutatingWebhookConfiguration{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+			requests := []reconcile.Request{}
+			if strings.HasPrefix(obj.Meta.GetName(), sidecarInjectorWebhookConfigNamePrefix) {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      sidecarInjectorServiceAccountSecretName,
+						Namespace: obj.Meta.GetName()[len(sidecarInjectorWebhookConfigNamePrefix):],
+					},
+				})
+			}
+			return requests
+		})})
+	if err != nil {
+		return err
+	}
+
+	// Watch ValidatingWebhookConfigurations
+	err = c.Watch(&source.Kind{Type: &v1beta1.ValidatingWebhookConfiguration{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+			requests := []reconcile.Request{}
+			if strings.HasPrefix(obj.Meta.GetName(), galleyWebhookConfigNamePrefix) {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      galleyServiceAccountSecretName,
+						Namespace: obj.Meta.GetName()[len(galleyWebhookConfigNamePrefix):],
+					},
+				})
+			}
+			return requests
+		})})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// reconciles webhook configurations
+type reconciler struct {
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// Reconcile reacts to changes in secrets by updating the corresponding webhook
+// configurations
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Secret", request.NamespacedName)
+	// get secret
+	secret := &corev1.Secret{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, secret)
+	if err != nil {
+		// the secret has since disappeared. Ignore
+		return reconcile.Result{}, nil
+	}
+	caRoot, ok := secret.Data[rootCertKey]
+	if !ok {
+		logger.Error(errors.New("secret does not contain root certificate"), "Incomplete secret")
+		return reconcile.Result{}, nil
+	}
+	updated := false
+	var config runtime.Object
+
+	if request.Name == sidecarInjectorServiceAccountSecretName {
```

Comment:
Would these work better as separate reconcilers, instead of if/else?  You could have one for mutating webhook and another for validating webhook, each would sync it's cabundle to the latest, so above, you'd be sending the webhook resources through the reconcile loop instead of the secrets.

Also, because you're setting the owner reference on the webhooks, you'll need to watch them for deletion and recreate them.  This could be tricky, as you need to know whether or not it's being deleted because of garbage collection or because the control plane is being deleted.  I think it would be easiest to remove the owner reference altogether, but this would block pod creation in member namespaces.  @knrc?

Reply:
I've been researching the ownerReferences stuff today, and it's still true that we're not using them correctly. Creating an ownerReference on a cluster-scoped object that points to a namespace-scoped owner is not supported. Only the inverse case is supported (see https://github.com/kubernetes/kubernetes/blob/6e5176b48df8719ab969a233a0411ad6ba35d222/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L288-L290)

Instead of the adding the ownerReference and letting the GC ensure that the webhook is deleted when the Deployment goes away, if the SMCP still exists, the Operator should re-create the Deployment anyway. Having the GC delete the webhook is unnecessary and just complicates things (the Operator should re-create the webhook in that case). 

Previously, the problem was if someone deleted the control plane namespace. But now, if that happens, the SMCP is deleted as well, so the Operator should (and probably already does) delete the webhook automatically.

Reply:
> I've been researching the ownerReferences stuff today, and it's still true that we're not using them correctly. Creating an ownerReference on a cluster-scoped object that points to a namespace-scoped owner is not supported. Only the inverse case is supported (see https://github.com/kubernetes/kubernetes/blob/6e5176b48df8719ab969a233a0411ad6ba35d222/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L288-L290)

That's what I was referring to on IRC yesterday. The docs clearly state that what we're doing is not supported, however it behaves as expected: webhooks are removed when their namespace-scoped owner (the deployment) is deleted. This 'wrong' use of ownerReferences is coming from upstream.

> Instead of the adding the ownerReference and letting the GC ensure that the webhook is deleted when the Deployment goes away, if the SMCP still exists, the Operator should re-create the Deployment anyway. Having the GC delete the webhook is unnecessary and just complicates things (the Operator should re-create the webhook in that case).
> 
> Previously, the problem was if someone deleted the control plane namespace. But now, if that happens, the SMCP is deleted as well, so the Operator should (and probably already does) delete the webhook automatically.

I'm bringing up a new cluster right now to test through these use cases, but I have created and deleted dozens of SMCPs in the last few days and never had a single leftover webhook, neither with or without the ownerReferences. That was also the reason it never occurred to me to add them. If we're running the operator we shouldn't rely on the GC for cleanup IMO. But I don't know if that's best practice.

Reply:
> Would these work better as separate reconcilers, instead of if/else? You could have one for mutating webhook and another for validating webhook, each would sync it's cabundle to the latest, so above, you'd be sending the webhook resources through the reconcile loop instead of the secrets.

I think that's a matter of taste. The biggest difference I see is that I would have to create two watches on `Secret`, one for each Webhook type. Possibly that is handled by controller-runtime internally by reusing the same watch. I don't know for sure though.

Reply:
An ownerRef pointing from a cluster-scoped dependent to a namespace-scoped owner is conceptually wrong, because it's not clear what namespace the owner is in. It might appear to work, but there are no guarantees that it will always work. A few months ago, someone reported the webhook disappearing after a few days. It turned out it was caused by a race condition in the GC, which only occurred when you restarted the "cluster" (more specifically, it happend during startup of the GC). One such race condition was fixed, but for the inverted ownerRef (from a namespace-scoped dependent to a cluster-scoped owner). Our use-case continues to be unsupported, so I consider it very dangerous, but I suggest we deal with this in a different PR. Your changes now keep the ownerRef as it was.

Reply:
Some history (it took me a long time to understand the purpose behind all this):

The ownerReference was used so that if the deployment pods disappeared, the webhook would also be removed.  This ensured that other deployments would succeed, even if the webhook pods were not available.  This was primarily and issue when we were watching all namespaces.  The side effect was that the GC would periodically delete the webhook, because it couldn't resolve the owner (no namespace on a webhook), so the deployments implementing the webhooks were modified to check for deletion and recreate them.

That said, the operator properly cleans up these resources, and, if we're deleting the webhooks when the deployments aren't there, we can simply change the failure policy on the webhooks so pod creation can continue if the webhook invocation fails.  This removes the need for ownerReference entirely, while preserving the existing behavior.  (You could also argue that since we're not watching all namespaces anymore, we could let pod creation hang in the member namespaces until the webhook service comes back.)

> >  Would these work better as separate reconcilers, instead of if/else? You could have one for mutating webhook and another for validating webhook, each would sync it's cabundle to the latest, so above, you'd be sending the webhook resources through the reconcile loop instead of the secrets.

> I think that's a matter of taste. The biggest difference I see is that I would have to create two watches on Secret, one for each Webhook type. Possibly that is handled by controller-runtime internally by reusing the same watch. I don't know for sure though.

That is correct, two reconcilers, each with its own watch.  My rationale was that we're actually reconciling the webhooks, not the secrets, so running reconcilation against the secrets inverts the logic, imo.